### PR TITLE
arch/stm32h7: add CM4 core support

### PR DESCRIPTION
**Pull request by raiden00pl**
## Summary

- armv7-m/mpu.h: add macro to configure shared memory region
- stm32h7: add HSEM support
- arch/stm32h7/rcc: default value for BOARD_FLASH_PROGDELAY
- arch/stm32h7: add an option to bypass clock configuration
- arch/stm32h7: use STM32_CPUCLK_FREQUENCY to initialize perf
- arch/stm32h7: add CM4 core support
- arch/stm32h7: add RPTUN support
- boards/stm32h7: add nucleo-h745zi
- boards: disable CM4 for stm32h745i-disco and stm32h747i-disco

## Impact
Boards that use H7 dual core chips but they don't care about CM4 should set `CONFIG_STM32H7_CORTEXM4_DISABLED=y`.
I keep this option disabled at default, because using dual core with CM4 disabled probably makes no sense.

Should have no impact on single core H7 but **configurations need refreshing**

## Testing
```
cmake -B build_h7m7 -DBOARD_CONFIG=nucleo-h745zi:nsh_cm7_rptun -GNinja
cmake -B build_h7m4 -DBOARD_CONFIG=nucleo-h745zi:nsh_cm4_rptun -GNinja
```
fix for not patched openamp issue:
```
rm openamp/.openamp_patch
cmake --build build_h7m4 -t menuconfig
```
```
cmake --build build_h7m7
cmake --build build_h7m4
```

flash CM4 and CM7 hex with `JFlashLiteExe`, device must be selected to STM32H745ZI_M7

### CODE REVIEW

File: arch/arm/src/stm32h7/hardware/stm32h7x3xx_memorymap.h [hunk 1]
```diff
@@ -60,10 +60,18 @@
 
 #define STM32_DTCRAM_BASE    0x20000000     /* 0x20000000-0x2001ffff: DTCM-RAM on TCM interface */
 #define STM32_AXISRAM_BASE   0x24000000     /* 0x24000000-0x247fffff: System AXI SRAM */
-#define STM32_SRAM1_BASE     0x30000000     /* 0x30000000-0x3001ffff: System SRAM1 */
-#define STM32_SRAM2_BASE     0x30020000     /* 0x30020000-0x3003ffff: System SRAM2 */
-#define STM32_SRAM3_BASE     0x3004c000     /* 0x30040000-0x30047fff: System SRAM3 */
-#define STM32_SRAM123_BASE   0x30000000     /* 0x30000000-0x30047fff: System SRAM123 */
+
+#ifdef CONFIG_ARCH_CHIP_STM32H7_CORTEXM7
```
Comment (davids5) [inline arch/arm/src/stm32h7/hardware/stm32h7x3xx_memorymap.h:64]: @raiden00pl Why is this here? Would you please explain this? Where are the second set of addresses from?

File: arch/arm/src/stm32h7/hardware/stm32h7x3xx_rcc.h [hunk 2]
```diff
@@ -812,6 +812,9 @@
 /* RCC Global Control register */
 
 #define RCC_GCR_WW1RSC                  (1 << 0)  /* Bit 0: WWDG1 reset scope control */
+#define RCC_GCR_WW2RSC                  (1 << 1)  /* Bit 1: WWDG2 reset scope control */
```
Comment (davids5) [inline arch/arm/src/stm32h7/hardware/stm32h7x3xx_rcc.h:815]: Is this defining reserved bit on other ICs? If so should we use and #ifdef on HAS_XXX to not allow compilations and to prevent errant access to reserved bits?

File: arch/arm/src/stm32h7/hardware/stm32h7x3xx_syscfg.h [hunk 3]
```diff
@@ -54,6 +54,7 @@
 
 #define STM32_SYSCFG_UR_OFFSET(n)     (0x0300 + ((n) << 2))
 #define STM32_SYSCFG_UR0_OFFSET       0x0300 /* User register 0 */
+#define STM32_SYSCFG_UR1_OFFSET       0x0304 /* User register 2 */
```
Comment (davids5) [inline arch/arm/src/stm32h7/hardware/stm32h7x3xx_syscfg.h:57]: Ditto

File: arch/arm/src/stm32h7/hardware/stm32h7x3xx_syscfg.h [hunk 4]
```diff
@@ -86,6 +87,24 @@
...
+#define STM32_SYSCFG_UR9              (STM32_SYSCFG_BASE + STM32_SYSCFG_UR9_OFFSET)
+#define STM32_SYSCFG_UR10             (STM32_SYSCFG_BASE + STM32_SYSCFG_UR10_OFFSET)
+#define STM32_SYSCFG_UR11             (STM32_SYSCFG_BASE + STM32_SYSCFG_UR11_OFFSET)
+#define STM32_SYSCFG_UR12             (STM32_SYSCFG_BASE + STM32_SYSCFG_UR12_OFFSET)
+#define STM32_SYSCFG_UR13             (STM32_SYSCFG_BASE + STM32_SYSCFG_UR13_OFFSET)
+#define STM32_SYSCFG_UR14             (STM32_SYSCFG_BASE + STM32_SYSCFG_UR14_OFFSET)
+#define STM32_SYSCFG_UR15             (STM32_SYSCFG_BASE + STM32_SYSCFG_UR15_OFFSET)
+#define STM32_SYSCFG_UR16             (STM32_SYSCFG_BASE + STM32_SYSCFG_UR16_OFFSET)
+#define STM32_SYSCFG_UR17             (STM32_SYSCFG_BASE + STM32_SYSCFG_UR17_OFFSET)
+
```
Comment (davids5) [inline arch/arm/src/stm32h7/hardware/stm32h7x3xx_syscfg.h:107]: Ditto

**Review by davids5 — COMMENTED** [review]
@raiden00pl Nice work. 

My comments are related to adding dual core code to non dual code CPUS and avoiding the stm32 issues of not bifurcating the code sooner.

I find CONFIG_ARCH_CHIP_STM32H7_CORTEXM7 confusing.  I am wondering if something like `CONFIG_ARCH_CHIP_STM32H7_DUAL_CORE_IS_M7` and  
`CONFIG_ARCH_CHIP_STM32H7_DUAL_CORE_IS_M4` would be better?

File: arch/arm/src/stm32h7/hardware/stm32h7x3xx_memorymap.h (hunk 1 above)
Comment (raiden00pl) [reply arch/arm/src/stm32h7/hardware/stm32h7x3xx_memorymap.h:64]: From RM0399 -> 2.4 Embedded SRAM:
```
The AHB SRAMs of the D2 domain are also aliased to an address range below
0x2000 0000 to maintain the Cortex®-M4 Harvard architecture:
•AHB SRAM1 also mapped at address 0x1000 0000 and accessible by all system
masters through D2 domain AHB matrix
•AHB SRAM2 also mapped at address 0x1002 0000 and accessible by all system
masters through D2 domain AHB matrix
•AHB SRAM3 also mapped at address 0x1004 0000 and accessible by all system
masters through D2 domain AHB matrix
```

File: arch/arm/src/stm32h7/hardware/stm32h7x3xx_syscfg.h (hunk 4 above)
Comment (raiden00pl) [reply arch/arm/src/stm32h7/hardware/stm32h7x3xx_syscfg.h]: URxx except UR1 are also available on stm32hx3xx

**Comment by raiden00pl** [comment]
> I find CONFIG_ARCH_CHIP_STM32H7_CORTEXM7 confusing. I am wondering if something like CONFIG_ARCH_CHIP_STM32H7_DUAL_CORE_IS_M7 and
CONFIG_ARCH_CHIP_STM32H7_DUAL_CORE_IS_M4 would be better?

I'm not sure about that. `CONFIG_ARCH_CHIP_STM32H7_DUAL_CORE_IS_M7` can also be confusing as we need that flag in both dual core and single core scenario. 
We can drop `DUAL_CORE` part and have `CONFIG_ARCH_CHIP_STM32H7_IS_M7` but I don't know if it's better than what it is now. `CORTEXM7`/`CORTEXM4` seems to be more obvious.
Maybe `CONFIG_ARCH_CHIP_STM32H7_IS_CORTEXM7` or `CONFIG_ARCH_CHIP_STM32H7_IS_CM7` ?

**Comment by davids5** [comment]
> > I find CONFIG_ARCH_CHIP_STM32H7_CORTEXM7 confusing. I am wondering if something like CONFIG_ARCH_CHIP_STM32H7_DUAL_CORE_IS_M7 and
> > CONFIG_ARCH_CHIP_STM32H7_DUAL_CORE_IS_M4 would be better?
> 
> @davids5 I'm not sure about that. `CONFIG_ARCH_CHIP_STM32H7_DUAL_CORE_IS_M7` can also be confusing as we need that flag in both dual core and single core scenario. We can drop `DUAL_CORE` part and have `CONFIG_ARCH_CHIP_STM32H7_IS_M7` but I don't know if it's better than what it is now. `CORTEXM7`/`CORTEXM4` seems to be more obvious. Maybe `CONFIG_ARCH_CHIP_STM32H7_IS_CORTEXM7` or `CONFIG_ARCH_CHIP_STM32H7_IS_CM7` ?
> 
> `CONFIG_ARCH_CHIP_STM32H7_CORTEXM7` is an evolution of `CONFIG_ARCH_CORTEXM7`

In my mind. I want to compartmentalize the dual core. So if that part number of the part is a dual core I see dual core files and settings. If it is not I do not.  I feel the naming of things that are specific to dual code, should indicate that. Hence adding something like `CONFIG_ARCH_CHIP_STM32H7_DUAL_CORE_IS_M7`  helps with that compartmentalization. The actual name could be more succinct.  But the idea is where there is an intersection in dual core/non dual core is is clear.

@raiden00pl  I am not as close to the problem or the solution space as you are, but my gut reaction is to duplicate the header files and remove some of the x in the naming to the fully defined STM32H745/755 and STM32H747/757 naming.

File: arch/arm/src/armv7-m/mpu.h [hunk 5]
```diff
@@ -339,26 +339,47 @@ void mpu_configure_region(uintptr_t base, size_t size,
...
+ ****************************************************************************/
+
+#define mpu_priv_shmem(base, size)                                      \
+  do                                                                    \
+    {                                                                   \
+      /* The configure the region */                                    \
+      mpu_configure_region(base, size,                                  \
+                           MPU_RASR_TEX_SO   | /* Ordered            */ \
+                                               /* Not Cacheable      */ \
+                                               /* Not Bufferable     */ \
```
[the remark this answers is not in the document]
Comment (raiden00pl) [reply arch/arm/src/armv7-m/mpu.h]: moved to https://github.com/apache/nuttx/pull/10344

**Comment by raiden00pl** [comment]
> @raiden00pl I am not as close to the problem or the solution space as you are, but my gut reaction is to duplicate the header files and remove some of the x in the naming to the fully defined STM32H745/755 and STM32H747/757 naming.

I don't think there is any need to duplicate the headers. The headers are the same for H7 chips except for literally a few extra bits when CM4 is supported.

**Review by davids5 — APPROVED** [review]
@raiden00pl - Thank you!

File: arch/arm/src/stm32h7/hardware/stm32h7x3xx_memorymap.h (hunk 1 above)
Comment (hartmannathan) [reply arch/arm/src/stm32h7/hardware/stm32h7x3xx_memorymap.h:64]: @raiden00pl is it possible to put that as a comment near the code? In the future, it will be easier to understand.

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/arm/src/stm32h7/Kconfig [hunk 6]
```diff
@@ -633,6 +633,10 @@ config STM32H7_LPTIM
 	bool
 	default n
 
+config STM32H7_HSEM
+	bool "Hardware sempahore"
```
Comment (acassis) [inline arch/arm/src/stm32h7/Kconfig:685]:
```suggestion
	bool "Hardware semaphore"

File: arch/arm/src/stm32h7/hardware/stm32_hsem.h [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+#define HSEM_SEMX_PROCID_SHIFT      (0)        /* Bits 0-8: Semaphore PROCID */
+#define HSEM_SEMX_PROCID_MASK       (0xff << HSEM_SEMX_PROCID_SHIFT)
+#define HSEM_SEMX_COREID_SHIFT      (8)        /* Bits 8-11: Semaphore COREID */
+#define HSEM_SEMX_COREID_MASK       (0xf << HSEM_SEMX_COREID_SHIFT)
+                                               /* Bits 12-30: reserved */
+#define HSEM_SEMX_LOCK              (1 << 31)  /* Bit 31: Lock indication */
+
+/* Clear register */
+
+#define HSEM_CR_COREID_SHIFT        (8) /* Bits 8-11: COREID of semaphores to be cleared*/
```
Comment (acassis) [inline arch/arm/src/stm32h7/hardware/stm32_hsem.h:74]:
```suggestion
#define HSEM_CR_COREID_SHIFT        (8) /* Bits 8-11: COREID of semaphores to be cleared */

File: arch/arm/src/stm32h7/hardware/stm32h7x3xx_memorymap.h (hunk 1 above)
Comment (raiden00pl) [reply arch/arm/src/stm32h7/hardware/stm32h7x3xx_memorymap.h:64]: done

**Comment by hartmannathan** [comment]
Reviewing...

File: arch/arm/src/stm32h7/Kconfig [hunk 8]
```diff
@@ -492,6 +503,39 @@ config STM32H7_AXI_SRAM_CORRUPTION_WAR
...
+if ARCH_STM32H7_DUALCORE
+
+if ARCH_CHIP_STM32H7_CORTEXM7
+
+config STM32H7_CORTEXM4_DISABLED
+	bool "Disable support for M4 core"
+	default n
+
+config STM32H7_CORTEXM7_BOOTM4
+	bool "Boot M4 core core"
```
Comment (hartmannathan) [inline arch/arm/src/stm32h7/Kconfig:515]: Just minor nit:

```suggestion
	bool "Boot M4 core"
```

File: arch/arm/src/stm32h7/Kconfig [hunk 9]
```diff
@@ -2086,6 +2138,15 @@ config STM32H7_QSPI_REGDEBUG
...
 
+config STM32H7_BYPASS_CLOCKCONFIG
+	bool "Bypass clock configuration"
+	depends on ARCH_STM32H7_DUALCORE
+	default n if ARCH_CHIP_STM32H7_CORTEXM7
+	default y if ARCH_CHIP_STM32H7_CORTEXM4
+	---help---
+		Bypass clock configuration. For dual core chips only one core
+		should configure clocks
+
```
Comment (hartmannathan) [inline arch/arm/src/stm32h7/Kconfig:2149]: I'm a little confused by `STM32H7_BYPASS_CLOCKCONFIG`: I understand that only one core should configure clocks, but which core will do that? Maybe the `---help---` text could clarify that a little bit.

File: arch/arm/src/stm32h7/stm32_allocateheap.c [hunk 10]
```diff
@@ -44,13 +44,21 @@
...
 #include "stm32_fmc.h"
 
 /****************************************************************************
  * Pre-processor Definitions
  ****************************************************************************/
 
+#if defined(CONFIG_ARCH_CHIP_STM32H7_CORTEXM7) && \
+    defined(CONFIG_STM32H7_CORTEXM4_DISABLED)
+
+/* Configuration for M7 core and M4 core support disabled */
```
Comment (hartmannathan) [inline arch/arm/src/stm32h7/stm32_allocateheap.c:60]:
```suggestion
/* Configuration for M7 core when M4 core support disabled */
```

I think that's the intended message?

**Review by hartmannathan — COMMENTED** [review]
I understand that the desire is to enable the Cortex M4 core by default, but allow the option to disable it at Kconfig-time. However, `CONFIG_STM32H7_CORTEXM4_DISABLED` causes many preprocessor conditionals to be in the form `#if !defined(CONFIG_STM32H7_CORTEXM4_DISABLED)` where the word "DISABLED" is very prominent and the `!` makes it a double-negative, so we see the word "DISABLED" but it really means "ENABLED"... Wouldn't it be better to call it CONFIG_STM32H7_CORTEXM4_ENABLED, enable it by default in Kconfig, and remove the double-negative from all the preprocessor conditionals? I know it's a lot of code churn for nothing, except (hopefully) being a little more clear in the future.

Also it might be nice to add some Documentation to clarify that NuttX needs to be built twice, once for each core. Right now, the only documentation of this fact is in the PR comments, and someone looking at the code in the future will have a difficult time figuring that out. This is already a pretty big PR so it's okay to add this kind of Documentation in a separate PR, but maybe add at least a comment or a hint somewhere, such as in the `---help---` text of ARCH_CHIP_STM32H7 or something?

**Comment by xiaoxiang781216** [comment]
@raiden00pl please rebase your patch to the last master, ci break is fixed.

File: arch/arm/src/stm32h7/Kconfig (hunk 8 above)
Comment (raiden00pl) [reply arch/arm/src/stm32h7/Kconfig]: fixed

File: arch/arm/src/stm32h7/stm32_allocateheap.c (hunk 10 above)
Comment (raiden00pl) [reply arch/arm/src/stm32h7/stm32_allocateheap.c]: correct :) fixed

File: arch/arm/src/stm32h7/stm32_allocateheap.c [hunk 11]
```diff
@@ -100,11 +108,33 @@
...
+#  define SRAM_START STM32_AXISRAM_BASE
+#  define SRAM_END   (SRAM_START + STM32H7_SRAM_SIZE)
+
+#  define SRAM123_START STM32_SRAM123_BASE
+#  define SRAM123_END   (SRAM123_START + STM32H7_SRAM123_SIZE)
+
+#elif defined(CONFIG_ARCH_CHIP_STM32H7_CORTEXM7) && \
+      !defined(CONFIG_STM32H7_CORTEXM4_DISABLED)
+
+/* Configuration for M7 core and M4 core support enabled */
```
[the remark this answers is not in the document]
Comment (raiden00pl) [reply arch/arm/src/stm32h7/stm32_allocateheap.c]: fixed

File: arch/arm/src/stm32h7/stm32_hsem.c [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+#else
+  putreg32((HSEM_COREID_CPU2 << HSEM_CR_COREID_SHIFT) , STM32_HSEM_RX(id));
+#endif
+}
+
+/****************************************************************************
+ * Name: stm32_hsem_init
+ *
+ * Description:
+ *   Intiialize the hardware semaphore driver
```
[the remark this answers is not in the document]
Comment (raiden00pl) [reply arch/arm/src/stm32h7/stm32_hsem.c]: fxied

File: arch/arm/src/stm32h7/stm32_hsem.c [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+{
+  struct stm32_hsem_s *dev    = args;
+  uint32_t             regval = 0;
+  int                  i      = 0;
+
+  regval = getreg32(STM32_HSEM_CXMISR(dev->block));
+
+  _info("HSEM interrupt 0x%" PRIx32 "\n", regval);
+
+  for (i = 0; i < STM32_HSEM_CHANS; i += 1)
```
[the remark this answers is not in the document]
Comment (raiden00pl) [reply arch/arm/src/stm32h7/stm32_hsem.c]: fixed

File: arch/arm/src/stm32h7/stm32_mpuinit.c [hunk 14]
```diff
@@ -42,18 +58,25 @@
  * Name: stm32_mpuinitialize
  *
  * Description:
- *   Configure the MPU to permit user-space access to only restricted SAM3U
- *   resources.
+ *   Configure the MPU.
+ *
+ *   If PROTETECTED build:
```
[the remark this answers is not in the document]
Comment (raiden00pl) [reply arch/arm/src/stm32h7/stm32_mpuinit.c]: fixed

File: boards/arm/stm32h7/nucleo-h745zi/include/board.h [hunk 15]
```diff
@@ -0,0 +1,393 @@
...
+
+/* Do not include STM32 H7 header files here */
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Clocking *****************************************************************/
+
+/* The Nucleo-144  board provides the following clock sources:
```
[the remark this answers is not in the document]
Comment (raiden00pl) [reply boards/arm/stm32h7/nucleo-h745zi/include/board.h:42]: I prefer to leave board files as they are now. This should be fixed for all H7 boards (and probably also other stm32 boards)

File: arch/arm/src/stm32h7/Kconfig (hunk 9 above)
Comment (raiden00pl) [reply arch/arm/src/stm32h7/Kconfig:2149]: This is the developer's decision. Both CPUs have RCC access. This is automatically selected so the dev doesn't have to do anything here. If someone wants to change the default behavior here, I assume they know exactly what they're doing

**Comment by raiden00pl** [comment]
> I understand that the desire is to enable the Cortex M4 core by default, but allow the option to disable it at Kconfig-time. However, CONFIG_STM32H7_CORTEXM4_DISABLED causes many preprocessor conditionals to be in the form #if !defined(CONFIG_STM32H7_CORTEXM4_DISABLED) where the word "DISABLED" is very prominent and the ! makes it a double-negative, so we see the word "DISABLED" but it really means "ENABLED"... Wouldn't it be better to call it CONFIG_STM32H7_CORTEXM4_ENABLED, enable it by default in Kconfig, and remove the double-negative from all the preprocessor conditionals? I know it's a lot of code churn for nothing, except (hopefully) being a little more clear in the future.

I agree, it makes sense. Changed.

> Also it might be nice to add some Documentation to clarify that NuttX needs to be built twice, once for each core. Right now, the only documentation of this fact is in the PR comments, and someone looking at the code in the future will have a difficult time figuring that out. This is already a pretty big PR so it's okay to add this kind of Documentation in a separate PR, but maybe add at least a comment or a hint somewhere, such as in the ---help--- text of ARCH_CHIP_STM32H7 or something?

doc will come later. First we have to solve https://github.com/apache/nuttx/issues/10241 to avoid some dirty hacks for build process. I assume this PR will only be used with cmake, compiling 2 images with make is too annoying.

File: arch/arm/src/stm32h7/Kconfig (hunk 9 above)
Comment (hartmannathan) [reply arch/arm/src/stm32h7/Kconfig:2150]: Thanks.

File: boards/arm/stm32h7/nucleo-h745zi/include/board.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (hartmannathan) [reply boards/arm/stm32h7/nucleo-h745zi/include/board.h:42]: No problems; thanks!

**Review by hartmannathan — APPROVED** [review]

**Comment by slorquet** [comment]
It would be good to avoid cmake-specific features.

Is make a supported build system or was it deprecated?

If it's supported no feature should work only with cmake. Thanks.

**Comment by xiaoxiang781216** [comment]
> It would be good to avoid cmake-specific features.
> 
> Is make a supported build system or was it deprecated?
> 
> If it's supported no feature should work only with cmake. Thanks.

The change should work with make too. But make doesn't support out of tree build, it's very annoying to develop AMP system like this patch because you have to make distclean and make every time.

**Comment by hartmannathan** [comment]
> > It would be good to avoid cmake-specific features.
> 
> > 
> 
> > Is make a supported build system or was it deprecated?
> 
> > 
> 
> > If it's supported no feature should work only with cmake. Thanks.
> 
> 
> 
> The change should work with make too. But make doesn't support out of tree build, it's very annoying to develop AMP system like this patch because you have to `make distclean` and `make` every time.

I started a discussion on the mailing list that is relevant to this: "[DISCUSS] Out-of-tree builds should be standard?" archived here: https://lists.apache.org/thread/l8mzkglx2dldx5jtxqpqtwmvyl4f4z8p

**Comment by raiden00pl** [comment]
@slorquet this PR includes make support, but there are some problems with OpenAMP build now. 

Without out of tree build it's a waste of time and without this specific cmake feature this PR would never have been created. I had dual core H7 in my plans for a long time, but the building process was too annoying (and sloooow) to take on it.

I have no problem modifying both cmake and make, unless support for make is too much work. 
There was a minimum of work here, but otherwise I'm not going to spend hours writing dedicated makefiles (unless someone pays me for this tedious job).

**Comment by raiden00pl** [comment]
UPDATE: this PR is not a breaking change, but the old configurations need refreshing. I wanted to avoid this, but Kconfig always add `CONFIG_ARCH_CHIP_STM32H7_CORTEXM7=y` to the refreshed configurations (probably the default behavior for `choice` options).

**Review by xiaoxiang781216 — APPROVED** [review]
